### DISCUSSION TITLE
Use ReplayGain (and Opus R128 Gain) tags for loudness normalization

### DISCUSSION THREAD

**Discussion by kepstin** [music-assistant/support#2872] [Feature requests and ideas]
As far as I can tell, Music Assistant currently uses ffmpeg's `loudnorm` filter to do loudness normalization. I'm unclear when that calculation is done (on demand? on import?), but it's a relatively cpu-heavy filter to run, and while the loudnorm filter's one-pass support for dynamic normalization (dynamic range compression, effectively) might be useful for radio or podcast streams, the filter is not always appropriate for local music files. (Note that for determining the integrated loudness of local files, ffmpeg's `ebur128` filter should run faster.)

But all of my music files contain pre-calculated "integrated" loudness adjustment information, stored either in the form of [ReplayGain 2.0 tags](https://wiki.hydrogenaud.io/index.php?title=ReplayGain_2.0_specification) or [Opus R128 Gain tags](https://www.rfc-editor.org/rfc/rfc7845.html#section-5.2.1). It would be great if Music Assistant could retrieve and use this precalculated loudness normalization information, rather than do its own loudness calculations. (Note that the ReplayGain 2.0 tags use a reference level of -18 LUFS, while the Opus R128 tags use a reference level of -23 LUFS.)

Ideally, the ability to select whether to use "track" or "album" normalization would also be provided - the "album" ReplayGain tags provide a single loudness adjustment for the album as a whole, while preserving dynamic differences between tracks within an album (this can be particularly important for concept albums, and ensures that tracks with gapless playback don't have loudness jumps when switching tracks).

**Comment by marcelveldt** [comment]
The loudnorm filter is not that heavy, at least other aspects of MA require more powerful hardware than that simple filter so I don't agree on that. Are you running it on a pi3 or even lower perhaps ?

Second of all most tags are just the simple replaygain tags to reduce the loudness of the entire track. Most often those calculations are based by a dumb "check the average loudness" algorithm instead of perreceived average loudness (EBU 128) causing issues with tracks with high dynamic range. It needs to be much, much smarter than that. I suggest you look into the loudnorm filter and the guy who wrote it and then realize that its miles better than just adjusting the gain of an entire track based on a single value. That is at least my opinion after comparing them both (we've had the old method that you propose here before btw).

We use loudnorm in 2 pass mode, which means that the first time you play something it runs in dynamic mode, basically collecting the measurement. It will not be as good at that point as its dynamically adjusting only with a limited look ahead. Now, once playback completes we store the measurement in the db (which is not a single value but a collection of values) and the very next time you play a track, it will then use those measured values to do the loudness leveling.

So I guess what I'm trying to say is that we implemented loudnorm for a reason; its way better than the common/dumb implementations of the replaygain tag. Maybe the Opus R128 tag is a bit better because that is at least the "average perceived loudness" EBU R128 calculation.

More background info about the loudnorm we choose to use in MA:
https://k.ylo.ph/2016/04/04/loudnorm.html

And a pointer to our implementation within the stream controller where you can also see what measurements values we store.
https://github.com/music-assistant/server/blob/main/music_assistant/server/controllers/streams.py#L798

**Reply by kepstin** [reply]
Most modern implementations of ReplayGain (including [the one I wrote](https://github.com/kepstin/regainer)) are based on the ReplayGain 2.0 specification, which uses the same EBU R.128 loudness measuring algorithm as used internally by loudnorm and used for the Opus tags. (ReplayGain 2.0 picked a reference level of -18 LUFS to approximately match the gain value calculated by the original less perceptually accurate ReplayGain algorithm).

I'd really prefer the loudness normalization I use for local files to:
* Preserve dynamics within a track - songs can have quieter and louder portions, but the current loudnorm setup limits the LRA of a track to ≤7 by dynamically applying gain adjustments if needed.
* Preserve dynamics within an album - albums can have quieter and louder tracks; the current loudnorm setup separately normalizes each track so they're all the same loudness.
* Apply a linear adjustment the first time I play a track, using values that were calculated in advance and stored in the file tags.

Note that on tracks with LRA ≤7 - which accounts for most modern pop/rock music - the single gain adjustment that ReplayGain provides will be the same as the adjustment applied by loudnorm's 2-pass linear mode. (Although you would need to add a peak limiter after applying the gain adjustment for the final result to be exactly the same.)

There certainly is a place for using loudnorm and per-track normalization on music! It is useful when you want to automatically reduce dynamic range when listening to music at lower volumes or in a noisy environment, where quiet sections might be inaudible otherwise. For this use case, an LRA even lower than 7 may be desirable.

**Reply by marcelveldt** [reply]
Isnt that exactly what our current implementation already does ?

**Reply by marcelveldt** [reply]
Also keep in mind that we have multiple sources and the main reason to use volume normalization within MA is when playing music from different sources, so for instance both Tidal and local files etc. I need to be absolutely sure how to interpret the existing measurement data in order to compare oranges with oranges.

**Reply by kepstin** [reply]
For ReplayGain 2.0 tags, the process is:
* Take the value of the `REPLAYGAIN_TRACK_GAIN` (or `REPLAYGAIN_ALBUM_GAIN`) tag, and parse it as floating point, discarding any trailing non-numeric characters. (e.g. `"-8.99 dB"` → `-8.99`) This value is the gain adjustment, in dB, required to adjust the playback level to -18 LUFS.
* Subtract that value from -18 to get the track's integrated loudness in LUFS. (e.g. `-18 - -8.99 = -9.01`) The resulting value is the same as the integrated loudness reported by the loudnorm filter first pass.

For Opus R128 tags, the process is:
* Take the value of the `R128_TRACK_GAIN` (or `R128_ALBUM_GAIN`) tag, and parse as integer. Divide by 256 to convert the value to floating point. (e.g. `"-3581"` → `-13.98828125`) This value is the gain adjustment, in dB, required adjust the playback level to -23 LUFS.
* Subtract that value from -23 to get the track's integrated loudness in LUFS: (e.g. `-23 - -13.98828125 = -9.01171875`) The resulting value is the same as the integrated loudness reported by the loudnorm filter first pass.

**Reply by marcelveldt** [reply]
Can you perhaps run a test on a few of your files if the measurement is indeed 1:1 the same as the loudnorm measurement ?
And loudnorm measures a few other things so I assume I cant use the loudnorm filter anymore based on single replaygain 2.0 / Opus R128 tags and I need to do a simple gain adjust ?

BTW: I see that your tool uses the ebur128 - we had that one before but it wasn't as good as loudnorm.
Most important because it meant that we needed to run a track through ffmpeg **before** we could play it, which is not always possible with streaming providers or makes it just super slow.

So to me it would be great if we can combine two worlds and use the correct filter at the correct time but still have confidence that tracks are equally normalized between sources.

So the value stored in the ReplayGain2.0/OpusR128 tag would be the Integrated loudness value but what about the other parameters loudnorm needs to do its job ? True peak, LRA and Treshold. 

I'm afraid we'll have to switch between filters based on the existence of the tag in the source material. It is kind of interesting though as some streaming providers do provide existing measurements as metadata. I simply wasn't able so far to map the different measurement techniques together with a proper filter that works as well as loudnorm now does.

That said, there are cases where loudnorm completely messes up, for example with very short audio clips or effect files like brown noise

**Reply by kepstin** [reply]
I've used ffmpeg's ebur128 and loudnorm filters quite a bit; the values that they report for Integrated Loudness and True Peak are always the same (which makes sense, since both are using the same shared library code to calculate the values).

Using the integrated loudness from the replaygain tags but feeding it into the second pass of the loudnorm filter is an idea I hadn't thought of before. It… kind of works, but has some downsides. ReplayGain does have tags to record the peak value, so that can be used as an input (the Opus R128 tags do not provide the peak value).

It looks like setting `measured_I=<replaygain loudness>:measured_TP=<replaygain peak if available, otherwise 0>:measured_LRA=<same as LRA>` and setting `measured_thresh` to … anything other than `-70` will put the loudnorm filter into linear mode when possible, giving the same behaviour as using a simple gain filter.

But if the condition `measured_TP + (I - measured_I) > TP` is true, then the loudnorm filter switches to dynamic mode in order to run the peak limiter. If this happens, the value of `measured_LRA` is ignored (dynamic range compression will be unconditionally applied), and the value of `measured_thresh` is used. If `measured_thresh` is set to a number close to -70, the results will be similar to loudnorm's 1-pass mode.

It's an interesting idea, but for applying replaygain values, I think it would be better in the end to use a different filter chain. This could be a simple gain filter (`volume=<I - measured_I>dB`), followed if necessary by a true peak limiter built from the `aresample` then … either the `alimiter` or `compand` filter could work, although finding good parameters might be tricky.

**Reply by kepstin** [reply]
If you're curious, a test track that I find works well for testing loudness normalization and peak limiting is Dire Straits - Money for Nothing off the original 1985 CD master of Brothers in Arms. Here's what the second pass of loudnorm using Music Assistant's settings (`loudnorm=I=-17:TP=-2.0:LRA=7.0:linear=true:print_format=summary:measured_I=-19.9:measured_TP=1.7:measured_LRA=12.2:measured_thresh=-30.4:offset=-0.5`) has to say about this one:

```
Input Integrated:    -19.9 LUFS
Input True Peak:      +1.7 dBTP
Input LRA:            12.2 LU
Input Threshold:     -30.4 LUFS

Output Integrated:   -17.2 LUFS
Output True Peak:     -2.0 dBTP
Output LRA:            6.9 LU
Output Threshold:    -27.4 LUFS

Normalization Type:   Dynamic
Target Offset:        +0.2 LU
```

So this track had overall loudness increased by 2.7 LU (dynamic mode undershot -17 LUFS a tiny bit), peaks cut by 6.4dB, more or less (hard to say with dynamic mode's variable gain), and loudness range (dynamic range) reduced from 12.2 LU to 6.9 LU.

That's not loudness normalization; that's basically a remaster of the song which completely changes the dynamics :( Probably makes it sound quite similar to the loudness wars era remasters of the track that are on the streaming services and CD re-releases nowadays.

One of the other tracks on this album has an Integrated loudness of -29.40 LUFS (almost 10 LU quieter!); when playing the album as a whole, it's somewhat jarring to have that be brought up to the same loudness as the louder tracks on the album.

**Reply by marcelveldt** [reply]
> That's not loudness normalization; that's basically a remaster of the song which completely changes the dynamics :( Probably makes it sound quite similar to the loudness wars era remasters of the track that are on the streaming services and CD re-releases nowadays.

Hmmmm, in dynamic mode loudnorm can bring some weird results but in linear mode its results should be satisfying. I listen to that track a lot and can not confirm on what you say here. Did you listen ? I see it ran in dynamic mode, despite the second pass, that is strange. 

I have no problem on switching to an alternative (I just have to browse back in history and find our previous implementation, its that simple) but I know that we switched for a reason; folks were not PERCEIVING the volume normalization and we got complaints if it was even working. For your reference; previously we just used the ebur128 filter to make a measurement of the overall loudness and then simply applied a gain shift on the total track - I didn't want to do normalization as that messes with dynamic range. Problem with the approach was playback of tracks without the measurement already in the db, we started guessing based on other tracks of the same album or even provider but that made little to no sense.

**Reply by marcelveldt** [reply]
> I've used ffmpeg's ebur128 and loudnorm filters quite a bit; the values that they report for Integrated Loudness and True Peak are always the same (which makes sense, since both are using the same shared library code to calculate the values).

OK that is a great start already then. We could then at least decide to keep the dynamic mode to make measurements on the fly maybe ?


> It's an interesting idea, but for applying replaygain values, I think it would be better in the end to use a different filter chain. This could be a simple gain filter (`volume=<I - measured_I>dB`), followed if necessary by a true peak limiter built from the `aresample` then … either the `alimiter` or `compand` filter could work, although finding good parameters might be tricky.

Yeah, that is exactly what we had before, just a simple gain filter, even without a limiter (and just make sure that we use sane values for target loudness) - use 24 bits or 32 bits bit depth internally and only dither that back to 16 bits if the player does not support 24 bits.

So, the idea is kind of intriguing; can we use loudnorm's dynamic filter if we have no measurement at all (and then let it collect the average perceived loudness accurately) but use existing measurements if we have them (and we are 100% sure what type of measurement it is) and then just go back to the simple gain/volume adjustment.

**Reply by marcelveldt** [reply]
I found our previous implementation in git history: https://github.com/music-assistant/server/blob/2.0.0b16/music_assistant/server/helpers/audio.py#L798

So that was basically 1 job to determine loudness using ebur128 filter and at playback time just a gain adjust using the volume filter.
As I said, that worked well with the exception of tracks that are not yet analyzed.

**Reply by kepstin** [reply]
> Hmmmm, in dynamic mode loudnorm can bring some weird results but in linear mode its results should be satisfying. I listen to that track a lot and can not confirm on what you say here. Did you listen ? I see it ran in dynamic mode, despite the second pass, that is strange.

The loudnorm filter second pass can't use linear mode, and falls back to dynamic mode, if either of the following conditions are true:
1. `measured_LRA` is greater than the `LRA` value you set, so automatic gain control is required to reduce the loudness range.
2. the required gain value (difference between `I` and `measured_I`) would cause the true peak to exceed the `TP` value you set, so peak limiting is required.

In the case of the Dire Straits track, **both** of these conditions are true. Its measured LRA of 12.2 exceeds the configured LRA of 7, and the measured true peak of 1.7 dbFS - combined with the a gain of +2.9 dB to bring it to the desired playback loudness - would exceed the configured true peak limit of -2 dbFS.

I listened to the result, and it's not unpleasant - but it's definitely different, if you're paying attention. The intro is made a fair bit louder, and the crescendo at around 1:30 that leads into the guitar riff isn't as impactful.

> Yeah, that is exactly what we had before, just a simple gain filter, even without a limiter (and just make sure that we use sane values for target loudness) - use 24 bits or 32 bits bit depth internally and only dither that back to 16 bits if the player does not support 24 bits.

Higher bit depth doesn't do anything to help clipping, although using a floating point format (which ffmpeg will do automatically in many cases) allows values above the max to be temporarily represented.

There's three ways to deal with potential clipping issues:
1. You can just let it clip when converting back to integer samples. This can introduce unpleasant distortion.
2. You can reduce gain to bring peaks below the limit. This will result in some songs playing back at lower than desired loudness levels.
3. You can use a limiter to attenuate peaks. Still introduces some distortion, but it's not as unpleasant or noticeable as clipping.

> So that was basically 1 job to determine loudness using ebur128 filter and at playback time just a gain adjust using the volume filter.

Yep - if you get lucky and have ReplayGain tags available, then you know someone has already gone ahead and determined the loudness in advance for you, so you just have to apply it :)

**Reply by marcelveldt** [reply]
The higher bit depth will help in heaving more headroom so you can preserve the dynamic range while doing the loudness adjustments.
Suggestion: We convert ALL incoming audio internally to PCM anyways so what we could consider is to internally use f64 - then do volume normalization (volume filter). Internally all audio is processed based on that pcm stream, including crossfading etc.

Then we have the final stage where we send the audio to the player, by default encoded to flac (because that is lossless) but it can also be in any other format. I think we should apply the limiter in this last chain then ?

Finally; we will use existing tags if they are present but if not, we will use the loudnorm dynamic mode as fallback (and also collect the measurement). For local files we could even go one step further and introduce a (slow) background task that pre-scans the files upfront.

How does that sound ? I do like to get your opinion on where to implement the limiter as I'm not entirely sure where in the chain that should be. This is the flow how we process all audio:

Stage 1: audio source --> decode with ffmpeg to PCM + loudnorm filter (replace by volume filter)
Stage 2: additional processing on the raw PCM if needed, such as crossfading, silence stripping etc.
Stage 3: encode to player-specific stream --> takes the raw PCM and encodes it to player specific format, takes care of downsampling if needed and apply any additional effects such as EQ, balance etc.

**Reply by kepstin** [reply]
I don't recommend using f64 (in ffmpeg, `dbl` or `dblp`) for audio, it's just a waste of memory. Using f32 (in ffmpeg, `flt` or `fltp`) provides enough precision - it's able to represent 24-bit integer audio exactly.

What sample rate is used for the intermediate audio processing stream? When `loudnorm` is operating in dynamic mode, I think it usually outputs 192kHz, but in linear mode it passes through the original sample rate. If peak limiting is being used, I recommend upsampling the audio to a higher sample rate before running the limiter.

If you allow audio effects like EQ, then it makes sense to have a peak limiter after the EQ (even when using loudnorm).

I like the idea of using loudnorm to dynamically adjust loudness and collect stats at the same time as a fallback. It might be interesting to experiment with the value of LRA. Larger values allow more dynamics - but also make it worse at adapting to shifts in loudness between tracks (or ads) on a radio stream.

Note that an option to force use of loudnorm (possibly with reduced or configurable LRA) could also useful - that can be used as a dynamic range reduction mode, like the "night mode" or "dynamic range control" feature some audio receivers have.

**Reply by marcelveldt** [reply]
f64 is a bit overkill indeed yeah, lets not so that. The intermediate sample rate is the original sample rate but sometimes upsampled (in case of flow mode streams, everything is resampled to the same). Yes loudnorm in dynamic mode does an intermediate upsample to 192khz and then outputs to our desired sample rate at the end of the chain.

I'll implement some of these ideas to play around with in a beta

**Reply by kepstin** [reply]
I did a bit of testing, and I found out that the filter chain `aresample=sample_rate=192K,alimiter=limit=-2dB:level=false:asc=true` works well as a true peak limiter.

On the `alimiter` filter, the `limit` param can be adjusted to taste (I used -2dB to match the settings being used with `loudnorm` right now). Note that when doing true peak limiting (with upsampling), running after EQ/effects, with floating point audio, the limit could be set as high as 0 dBFS - but using −2 dBFS gives extra headroom which might prevent clipping in processing done on low quality playback devices or caused by lossy audio encoding (mp3).

Setting `level=false` is required so that it does not perform any gain adjustments (by default it applies a gain opposite the limit in order to bring the peaks back up to 0 dbFS).

The `asc=true` option enables a feature where when detecting multiple peaks in short succession, it averages out the adjustment a bit rather than repeatedly applying the filter attack/release (the `loudnorm` peak limiter has a similar feature).

**Reply by marcelveldt** [reply]
Nice! Thanks for testing that. So this true peak limiter needs to be implemented at the end of chain right ? So in stage 2 of our internal processing where the pcm stream will be encoded to the player's preferences.

Only thing I wonder is if we really should upsample all material internally to 192K - I assume that is to correctly detect the true peaks.

**Reply by kepstin** [reply]
Yeah - if you want the `alimiter` filter to accurately/precisely limit the "true peak", then you need to upsample the audio.

But I did some additional testing, and the specific `alimiter` filter I recommend with `limit=-2dB` works well enough on 44.1 kHz and 48 kHz audio to keep the true peak below 0 dBFS in almost all cases (resulting true peak values around -0.6 dBFS to -1.8 dBFS are typical), which should be good enough to prevent any distortion due to clipping.

Note that "almost all" is not "all". I can find some counter-examples, e.g. one track with a true peak of 2.4 dBFS that only gets true peak reduced to 0.9 dBFS when running alimiter at 44.1kHz. That said, this track was fairly loud at -8.40 LUFS, so if volume normalization is applied the peaks get brought down well below -2 dbFS and the peak limiter doesn't need to do anything.

**Reply by marcelveldt** [reply]
Great thanks for testing and for being such a valuable source of information once again. BTW: Good to see you here as well, next to Matter/Thread ! I've already started working on something to enhance the volume normalization logic. I'm currently at IFA Berlin on behalf of Home Assistant / Nabu Casa but I'll probably finish this tomorrow in the train back  to home and release a beta to play with.

**Comment by GreenhouseAU** [comment]
So I can't really tell from the replies, but is the 'implemented' tag correct? (if so, how do I enable? :) )

All of my music has replaygain 2.0 tags, and I don't like the idea of listening to thousands of tracks before a 'loudness' value is assigned to a track, when you can do all the processing beforehand and have it stored in a platform-independent manner (ID3 or FLAC metadata).

Also, most of my listening is of albums, so all I need is one 'normalisation' value per album (i.e. album gain tag).

**Reply by OzGav** [reply]
The original request has been implemented. It works automatically. Look at the audio pipeline to see what volume normalisation is being applied during playback

**Reply by GreenhouseAU** [reply]
Ah, I have found the issue. 

With Plex (what I'm using because local disk doesn't play nicely with my folder structure/album art), the replaygain tags musn't be passed through to MA.

Files on local disk work well with replaygain.

So it appears I can have album art or replaygain, but not both :P

**Reply by OzGav** [reply]
Don’t forget option c is to have both with an adjustment to your folder structure (not sure what the problem there is).

**Comment by galettof** [comment]
Question: if I add ReplayGain tags to my music files after they have already been played in Music Assistant, will MA pick up the new tags or will it keep using the normalization values it previously computed? If MA keeps using its own stored values, how can I clear them and force MA to read my ReplayGain tags? Thanks.

**Reply by marcelveldt** [reply]
The measurement in the file is preferred and it will overwrite any stored value in the db

**Reply by galettof** [reply]
Thanks for the response. I've been paying closer attention during playback and I observed the following: the ReplayGain tags (album and track) in my files show certain values -x dB and -y dB, but the audio pipeline view in MA is showing a third different value -z dB. Is this to be expected or does this mean MA is not picking up the ReplayGain tags?

**Reply by OzGav** [reply]
How different is it. What does the more info tooltip show in the audio pipeline?
<img width="543" height="532" alt="image" src="https://github.com/user-attachments/assets/dc3f93cd-980f-496a-a155-04c15314bea4" />

**Reply by galettof** [reply]
Aha! The difference (z-y) is always 1.03 or 1.04. Aside from rounding, I am guessing this is due to the program I used to create the ReplayGain tags (rsgain) targeting -18 LUFS, while MA targets -17 LUFS. Does that seem reasonable? If so, it should be fine; my main goal was to ensure consistent adjustment within albums.

**Reply by OzGav** [reply]
I’m no expert on this but does the replaygain tag value get shown in the tooltip

**Reply by galettof** [reply]
No, it doesn't. I attached a screenshot. It says volume normalization is using a stored value, hence my initial confusion (was it using the value MA computed and stored, or the ReplayGain tag stored in the file?). The ReplayGain album tag is -9.88 dB and the track tag is -10.02 dB.

<img width="480" height="274" alt="01" src="https://github.com/user-attachments/assets/97cbdd7b-7504-487f-8732-dfac55e40951" />

**Reply by galettof** [reply]
I just tried another album which I had never previously played in MA. I used rsgain to compute ReplayGain tags *before* playing the album for the first time in MA. The tooltip showed MA was using the dynamic normalizer which, according to the documentation, is used when no integrated loudness measurement is available. This suggests to me that MA is not reading the ReplayGain tags, unless I misunderstand what is supposed to happen here (which is very possible).

I am using MA 2.7.0b6 installed as an add-on in Home Assistant OS with Core version 2025.10.4. The files are FLAC and are served to MA by a Jellyfin server. The player uses DLNA.

**Reply by marcelveldt** [reply]
There is a chance that the measurements are not properly passed by Jellyfin. As a test, you could perhaps try the builtin local files provider, if only with a small set of files to test

**Reply by galettof** [reply]
Thanks! I can give it a try.

**Reply by kepstin** [reply]
> Aha! The difference (z-y) is always 1.03 or 1.04. Aside from rounding, I am guessing this is due to the program I used to create the ReplayGain tags (rsgain) targeting -18 LUFS, while MA targets -17 LUFS. 

Indeed, the [ReplayGain 2.0 spec](https://wiki.hydrogenaudio.org/index.php?title=ReplayGain_2.0_specification) uses a reference level of -18 LUFS, and Music Assistant uses that as the basis for calculating the measured loudness of the track from the ReplayGain tags. Note that Music Assistant's default normalization to -17 LUFS is configurable per-player in the player advanced settings; you can change it to -18 to match ReplayGain levels if you prefer, or any other value you like to match other audio source levels if needed.

The method Music Assistant uses for measuring loudness only provides one decimal of precision; other tools may provide two decimal places of precision, so seeing a difference of <0.05 dB is pretty normal, and won't be audible.

**Reply by galettof** [reply]
Thanks for the explanation! The reason I've been looking into this is that on first pass, when the dynamic normalizer is used, I got the impression that MA may play different tracks on the same album at different loudness levels. This appears to be resolved by the second read when the loudness level is stored, and presumably would be resolved on first read if the existing loudness tags were passed correctly by Jellyfin. At that point -18 or -17 LUFS isn't a big deal for me, but good to know that the value can be configured.

**Reply by galettof** [reply]
Indeed, the ReplayGain tags are read correctly from local files, so it looks like Jellyfin is not passing that information.

**Reply by kepstin** [reply]
> on first pass, when the dynamic normalizer is used, I got the impression that MA may play different tracks on the same album at different loudness levels.

The dynamic normalizer has to work from scratch on each track. It picks the initial adjustment based on just the first few seconds of the track, then adjusts up or down to try to get the loudness to the expected level as song dynamics change. This means that if you have a track which starts quietly, it will sound "louder" than the previous track (at least for a bit, until the normalization brings it down - which in many cases is done gradually so it doesn't disrupt the music too much)

**Reply by galettof** [reply]
Thanks, that makes sense.

**Comment by rightsaidfred99** [comment]
If loudnorm is so good...why do I get WILD volume differences between songs? 

Who cares if ReplayGain is old when it is much more consistent inbetween tracks.

**Reply by marcelveldt** [reply]
And you confirmed in the output details that it is not using the dynamic limiter?

**Reply by kepstin** [reply]
There's a little secret... If you're playing from local files via the Filesystem (local disk or remote share) provider, Music Assistant will use the ReplayGain tags from your files as-is.

The loudnorm filter (dynamic limiter)... isn't great for individual tracks, if I'm gonna be honest. But it's usually better than not having any volume adjustment at all, and Music Assistant has a neat trick! After playing a track using the loudnorm filter once, it will remember the integrated loudness measurement that the loudnorm filter output when the track finished playing (equivalent to the "track ReplayGain" value) and will use that to apply a simple volume adjustment **next** time you play the track.

**Comment by de666** [comment]
Hello everyone I try to summarize my case/problem that I am trying to understand how best to solve.
My situation is as follows: Navidrome as a provider Audio files in various formats, both lossy and lossless, including opus.
After reading a bit of information I "discovered" that foobar2000 which is the software I use to manage files, mainly to convert from lossless to lossy and to rename and move the files, applies a sort of correction to the gain info for the opus format (R128_replaygain) by adding 5db, as explained [here](https://github.com/complexlogic/rsgain?tab=readme-ov-file#supported-file-formats) in the Opus section.
The problem is that the behavior of foobar2000 is different (obviously) even when it handles reading information.
So, referring to a playlist of mixed types of files (both opus and others) with replaygain info analyzed and tagged with foobar2000,
this is played with the same volume level if you use foobar2000 while, for example, on Music Assistant, opus files are played with an extremely lower volume (consistent with the 5db offset mentioned above).
This behavior leads me to think (tell me if I'm wrong) that Music Assistant manages the two types of tags in different ways, using a normalization to -23 LUFS for opus tags (R128_replaygain) and -18 LUFS for other file types (standard replaygain tags). Can you confirm that this is the case?
If this is confirmed, I don't know how to manage the opus tags to make sure that I can play the files at the same volume with both foobar2000 and with music assistant. Any ideas about it?
Also I still have doubts on how the two types of tags are handled by music assistant if they are both present on an opus file.

**Reply by OzGav** [reply]
This is a bug in the subsonic provider which was fixed today and will be rolled out in the upcomng updates

**Comment by rightsaidfred99** [comment]
@OzGav , thank you so much. Your fix works perfectly now :D

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
